src/main/java/com/czertainly/core/service/cmp/impl/CmpServiceImpl.java

```diff
@@ -0,0 +1,344 @@
+                        PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            } else {
+                LOG.error("{} | handling failed: \n\nrequest:\n {}\n response:\n {}", logPrefix,
+                        requestAsString, PkiMessageDumper.dumpPkiMessage(pkiResponse), e);
+            }
+            return buildBadRequest(pkiResponse);
+        }
+    }
+
+    private ResponseEntity buildBadRequest(PKIMessage pkiMessage) {
+        return ResponseEntity
+                .status(HttpStatus.BAD_REQUEST)
+                .header("Content-Type", HTTP_HEADER_CONTENT_TYPE)
+                .body(PkiMessageError.encode(pkiMessage));
+    }
+
+    private ResponseEntity buildOk(PKIMessage pkiMessage) throws IOException {
+        return ResponseEntity
+                .status(HttpStatus.OK)
+                .header("Content-Type", HTTP_HEADER_CONTENT_TYPE)
+                .body(pkiMessage.getEncoded());
+    }
+
+    private void init(String profileName) throws CmpConfigurationException {
+        this.raProfileBased = ServletUriComponentsBuilder.fromCurrentRequestUri().build()
+                .toUriString().contains("/raProfile/");
+        if (raProfileBased) {
+            raProfile = raProfileRepository.findByName(profileName).orElse(null);
+            if (raProfile == null) {
+                return;
+            }
+            cmpProfile = raProfile.getCmpProfile();
+            String attributesJson = raProfile.getProtocolAttribute() != null ? raProfile.getProtocolAttribute().getCmpIssueCertificateAttributes() : null;
+            issueAttributes = AttributeDefinitionUtils.getClientAttributes(AttributeDefinitionUtils.deserialize(attributesJson, DataAttribute.class));
+            String revokeAttributesJson = raProfile.getProtocolAttribute() != null ? raProfile.getProtocolAttribute().getCmpRevokeCertificateAttributes() : null;
+            revokeAttributes = AttributeDefinitionUtils.getClientAttributes(AttributeDefinitionUtils.deserialize(revokeAttributesJson, DataAttribute.class));
+        } else {
+            cmpProfile = cmpProfileRepository.findByName(profileName).orElse(null);
+            if (cmpProfile == null) {
+                return;
+            }
+            raProfile = cmpProfile.getRaProfile();
+            if (raProfile == null) {
+                return;
+            }
+            issueAttributes = attributeEngine.getRequestObjectDataAttributesContent(cmpProfile.getRaProfile().getAuthorityInstanceReference().getConnectorUuid(), AttributeOperation.CERTIFICATE_ISSUE, Resource.CMP_PROFILE, cmpProfile.getUuid());
+            revokeAttributes = attributeEngine.getRequestObjectDataAttributesContent(cmpProfile.getRaProfile().getAuthorityInstanceReference().getConnectorUuid(), AttributeOperation.CERTIFICATE_REVOKE, Resource.CMP_PROFILE, cmpProfile.getUuid());
+        }
+
+        Certificate cmpCaCertificate = cmpProfile.getSigningCertificate();

```

Comment:
This is also validated in `validateCmpProfile` method. Isn't it duplicate?

Reply:
rESOLVED